### ISSUE TITLE
[Feature Request] Custom FPS

### ISSUE THREAD

**Issue by Weeppiko** [nesbox/TIC-80#1589]
The solution mentioned in #294 it's not good because it doesn't take into account the fact that everything else still updating at 60hz. Since you're actually dropping frames with this solution, API functions like btnp() and keyp() could be missed in one of the dropped frames, the audio will still crack if the frame takes more than 16ms to end, etc. Be able to set a custom FPS (or at least an option for 30 FPS) that change the rate of the entire API would be very useful, specially for heavy games that make use of 3D visuals or raycast.

**Comment by RobLoach** [comment]
Since it's 60 FPS, you could update/render a frame every second frame to virtually achieve 30 FPS.

**Comment by Weeppiko** [comment]
But the audio cracks will still being an issue since it still being update at 60hz. The purpose of decrease the FPS on heavier games it's because they are take too long to end a frame, so just dropping every second frame will not help, because the frames that will be rendered will still take too long, and the audio will cracks on that frames.

**Comment by joshgoebel** [comment]
> But the audio cracks will still being an issue since it still being update at 60hz.

I don't see how.

>  The purpose of decrease the FPS on heavier games it's because they are take too long to end a frame

Now that is a indeed different problem if your script is simply hogging the CPU too long... I'd try to instead focus on what is making it so slow rather than wish for 30 FPS.

**Comment by Weeppiko** [comment]
Ok, listen to the actual situation:
I'm trying to use a raycast system to render the graphics, my current script runs like 56~48 FPS. Since you can modify the audio registers at runtime, the audio updates at the same rate as the game, if the game lags the audio also lags. A good solution for this would be to run the game at 30 FPS, so I tried to do the solution already mentioned, my game consistently runs at 30 FPS, but it doesn't help with the audio issue, the crack stills happens.

**Comment by joshgoebel** [comment]
You need to pretend it's real hardware - that there are no magic upgrades coming.  As @RobLoach said here probably are not. If you truly can't speed it up... then I would architect the game to update at 30 FPS but ALSO finish each TIC in the allotted 16ms... so you may need to raycast half the screen in one TIC and the other half in the second TIC so that your render is 30 FPS but your TICs are an *even* 60 FPS.

**Comment by Weeppiko** [comment]
If it's to pretend that it's a real hardware, then TIC-80 needs to behave more like an old console, when the game lags the audio generally doesn't cracks like TIC-80 does, but just slowdown it by keeping playing the last frequency that was set on that audio register. Also thank you for point me to a solution, I think it don't even need the use of coroutines at all.

**Comment by joshgoebel** [comment]
> when the game lags the audio generally doesn't cracks like TIC-80 does, but just slowdown it by keeping playing the last frequency that was set on that audio register.

I'm guessing it's not multi-threaded... the audio queue needs CPU time to be filled... so if your code is still running in the Lua/Wren/whatever VM then there is no time to fill then audio queue and it empties out.  I suppose you could maybe have a larger queue and overfill it (with extra samples) and then constantly reset it (no idea if that's actually possible with SDL2 - or if that would make the audio worse) but then you'd just be punishing all the games that were well-behaved to not burn too much CPU time.

The simplest way to have clean audio though is to have enough CPU time to fill the buffer - not to overfill it constantly.

**Comment by joshgoebel** [comment]
>  I think it don't even need the use of coroutines at all.

Yeah if your renderer is easily split you might not need them, but if you have a bunch of state sometimes it's easier to just yield in the middle of that work (preserving state) than to split the work...
